lfs/transfer_queue.go

```diff
@@ -278,6 +306,21 @@ func (q *TransferQueue) run() {
 	}
 }
 
+func (q *TransferQueue) canRetry(err error, id string) bool {
+	if !IsRetriableError(err) {
+		return false
+	}
+
+	defer q.retrylock.Unlock()
+	q.retrylock.Lock()
```

Comment:
As an alternative to this lock & map, how about:
```diff
diff --git a/lfs/download_queue.go b/lfs/download_queue.go
index d8e00e2..5c1b0f4 100644
--- a/lfs/download_queue.go
+++ b/lfs/download_queue.go
@@ -2,6 +2,7 @@ package lfs
 
 // The ability to check that a file can be downloaded
 type DownloadCheckable struct {
+	*TransferableBase
 	Pointer *WrappedPointer
 	object  *objectResource
 }
@@ -15,6 +16,7 @@ func (d *DownloadCheckable) Check() (*objectResource, error) {
 }
 
 func (d *DownloadCheckable) Transfer(cb CopyCallback) error {
+	d.RecordAttempt()
 	// just report completion of check but don't do anything
 	cb(d.Size(), d.Size(), int(d.Size()))
 	return nil
@@ -58,6 +60,7 @@ func NewDownloadable(p *WrappedPointer) *Downloadable {
 }
 
 func (d *Downloadable) Transfer(cb CopyCallback) error {
+	d.RecordAttempt()
 	err := PointerSmudgeObject(d.Pointer.Pointer, d.object, cb)
 	if err != nil {
 		return Error(err)
diff --git a/lfs/transfer_queue.go b/lfs/transfer_queue.go
index 636fb2a..30d9992 100644
--- a/lfs/transfer_queue.go
+++ b/lfs/transfer_queue.go
@@ -20,6 +20,18 @@ type Transferable interface {
 	Size() int64
 	Name() string
 	SetObject(*objectResource)
+	AttemptCount() int
+}
+
+type TransferableBase struct {
+	attempts int
+}
+
+func (t *TransferableBase) AttemptCount() int {
+	return t.attempts
+}
+func (t *TransferableBase) RecordAttempt() {
+	t.attempts++
 }
 
 // TransferQueue provides a queue that will allow concurrent transfers.
diff --git a/lfs/upload_queue.go b/lfs/upload_queue.go
index ce06545..649952a 100644
--- a/lfs/upload_queue.go
+++ b/lfs/upload_queue.go
@@ -8,6 +8,7 @@ import (
 
 // Uploadable describes a file that can be uploaded.
 type Uploadable struct {
+	*TransferableBase
 	oid      string
 	OidPath  string
 	Filename string
@@ -42,6 +43,7 @@ func (u *Uploadable) Check() (*objectResource, error) {
 }
 
 func (u *Uploadable) Transfer(cb CopyCallback) error {
+	d.RecordAttempt()
 	wcb := func(total, read int64, current int) error {
 		cb(total, read, current)
 		return nil
```
... and then pass the actual Transferable to this function and retry if t.AttemptCount() < 2.